src/ralph/backends/lrs/fs.py

```diff
@@ -0,0 +1,348 @@
+                    return True
+
+            statement_object = statement.get("object", {})
+            if statement_object.get("objectType") == "SubStatement":
+                return match_related_account(statement_object)
+            return False
+
+        if name or home_page:
+            filters.append(match_related_account if related else match_account)
+
+    @staticmethod
+    def _add_filter_by_verb(filters: list, verb_id: Union[str, None]) -> None:
+        """Add the `match_verb_id` filter if `verb_id` is set."""
+
+        def match_verb_id(statement: dict) -> bool:
+            """Return `True` if the statement has the given `verb.id`."""
+            return statement.get("verb", {}).get("id") == verb_id
+
+        if verb_id:
+            filters.append(match_verb_id)
+
+    @staticmethod
+    def _add_filter_by_activity(
+        filters: list, object_id: Union[str, None], related: Union[bool, None]
+    ) -> None:
+        """Add the `match_object_id` filter if `object_id` is set."""
+
+        def match_object_id(statement: dict) -> bool:
+            """Return `True` if the statement has the given `object.id`."""
+            return statement.get("object", {}).get("id") == object_id
+
+        def match_related_object_id(statement: dict) -> bool:
+            """Return `True` if the statement has any object.id matching `object_id`."""
+            statement_object = statement.get("object", {})
+            if statement_object.get("id") == object_id:
+                return True
+            activities = statement.get("context", {}).get("contextActivities", {})
+            for activity in activities.values():
+                if isinstance(activity, dict):
+                    if activity.get("id") == object_id:
+                        return True
+                else:
+                    for sub_activity in activity:
+                        if sub_activity.get("id") == object_id:
+                            return True
+            if statement_object.get("objectType") == "SubStatement":
+                return match_related_object_id(statement_object)
+
+            return False
+
+        if object_id:
+            filters.append(match_related_object_id if related else match_object_id)
+
+    @staticmethod
+    def _add_filter_by_timestamp_since(
+        filters: list, timestamp: Union[datetime, None]
+    ) -> None:
+        """Add the `match_since` filter if `timestamp` is set."""
+
+        def match_since(statement: dict) -> bool:
+            """Return `True` if the statement was created after `timestamp`."""
+            try:
+                statement_timestamp = datetime.fromisoformat(statement.get("timestamp"))
+            except (TypeError, ValueError) as error:
+                msg = "Statement with id=%s contains unparsable timestamp=%s"
+                logger.debug(msg, statement.get("id"), error)
+                return False
+            return statement_timestamp > timestamp
```

Comment:
Quick question concerning all backends: do you know why we don't return, for a `since` request, timestamp that are equals? And only for the `until` parameter

Reply:
Good question) We follow the [xAPI spec](https://github.com/adlnet/xAPI-Spec/blob/master/xAPI-Communication.md#user-content-2.1.3.s1.table1.row9) in this case) since is exclusive and until inclusive.

Reply:
Oh alright thanks!